791-kpi-data-export-specification

eatyourpeas opened the pull request:
### Overview

Implements request from E12 team to download KPIs
This PR does not implement all the requests from E12 team. In particular they have requested quarterly cohort reports which currently is not implemented.

This PR therefore adds a new template for downloading, with a new download button, which returns a single sheet CSV with summary data for each measure for the currently submitting cohort.

### Code changes

2 new URLs to `urls.py` - one for the new page, one for the csv download
2 new functions for each url in `organisation_views.py`
a new template for the new page
a new button to the `selected_organisation_summary` template
an opportunistic fix for permissions checks in `epilepsy12_user_list` template.
a new function to query the NationalKPIAggregation table against cohort 6, remap the fields into a pandas dataframe and return this for csv conversion and download in the view

### Documentation changes (done or required as a result of this PR)
No documentation

### Related Issues
Partially addresses issues in #791

### Mentions
Builds on work started by @pacharanero and @dc2007git

dc2007git committed:
Adds download KPI button to organisation summary page

dc2007git committed:
Fix to has_group filter for e12 team member calls

dc2007git committed:
Creates KPI download webpage and associated views

eatyourpeas committed:
refactor url to single kpi_download endpoint
refactor kpi_download function to return htmx on button press
replace download button with htmx partial label on click
include back button in kpi_download page

eatyourpeas committed:
query for national aggregations

eatyourpeas committed:
updates seed_registrations to calculate KPIs on complete
adds csv download endpoint
maps national KPI aggregation fields to measures and calculates percentages

eatyourpeas committed:
add decorator back to download function

mbarton commented on epilepsy12/views/organisation_views.py:
```diff
@@ -427,3 +432,29 @@ def view_preference(request, organisation_id, template_name):
+
+
+@login_and_otp_required()
+@permission_required("epilepsy12.can_publish_epilepsy12_data")
```
[from code review] this is equivalent to the RCPCH team (superuser). We're re-using the permission - we do have an `isRCPCH` equivalent in the user model but there isn't a permission associated with it and it feels heavy to add it for this PR

eatyourpeas replied:
perhaps just me but I am not sure that it is inappropriate to use this permission for this use case here. The point of E12 staff downloading the data is for the purposes of publication ultimately, and often the data analysis they will be doing here may be a precursor to formal publication, so to my mind this is part of the same thing. I am not sure what adding a new permission would gain, and it would create a new migration also.

pacharanero replied:
I agree it's acceptable for now, we just flagged the permission naming as a potential issue for the future. I think it's a stretch for most people to see that downloading data is 'eventually equivalent' to publication, because really there is no guarantee that the tweo activities are linked. Some downloaded data might be for other purposes.

If we find we are 'overloading' this permission in the future, we could consider either:
* a new permission specifically for downloading of E12 data like `can_download_epilepsy12_data`
* a broader permission, something like `is_epilepsy12_team`

At this stage fine as it is, but this is the kind of stuff that makes maintainability harder for other people less familiar with the codebase, so it's a thing to keep tabs on.

dc2007git committed:
Updates National measures to be in line with E12 template

pacharanero commented on epilepsy12/tasks.py:
```diff
@@ -69,10 +73,114 @@ def async_insert_old_pt_data(csv_path: str = "data.csv"):
+    national_kpi_aggregation = (
+        NationalKPIAggregation.objects.filter(cohort=cohort)
+        .values(
+            "paediatrician_with_expertise_in_epilepsies_passed",
```
@eatyourpeas is there a reason we are specifying each and every attribute here? From my understanding if you call `.values()` without attribute names you get them all, as a dict. We may get a couple we don't actually need, but this is probably acceptable I think. From my POV `national_kpi_aggregation = NationalKPIAggregation.objects.filter(cohort=cohort).values().first()` feels cleaner.

eatyourpeas replied:
Happy to go with this.

dc2007git committed:
Creates Country level csv dataframes. Imports openpyxl for eventual excelwriter processing

dc2007git committed:
Returns dataframe for Trust KPI aggregations

dc2007git committed:
returns trust AND health boards (wales) dataframe

dc2007git committed:
Returns ICB dataframe

dc2007git committed:
Returns Region dataframe, reorders column order as per template

dc2007git committed:
returns network dataframe

dc2007git committed:
Tidies up code with DRY principle - function written to create dataframes for each sheet

dc2007git committed:
Creates .xlsx document (without reference and national_comparison), also without health boards in NHSregion page

dc2007git committed:
Update exported excel to include Health Boards as an NHS region

dc2007git committed:
Update gitignore to ignore KPI export .xlsx doc

mbarton commented on epilepsy12/views/organisation_views.py:
```diff
@@ -1,11 +1,15 @@
 # Python imports
 from datetime import date
+import asyncio
```
Is this used anywhere?

eatyourpeas replied:
no sorry. I think added in error. Please do remove.

mbarton commented on epilepsy12/tasks.py:
```diff
@@ -69,10 +73,288 @@ def async_insert_old_pt_data(csv_path: str = "data.csv"):
+        "school_individual_healthcare_plan",
+    ]
+
+    def create_dataframe(KPI_model1, constants_list, KPI_model2=None, constants_list2=None, is_regional=False):
```
Would it make sense to move this function out to be at the top level somewhere?

At the moment it's hard to see at a glance where it ends, although there is the `# COUNTRY - SHEET 1` comment

dc2007git replied:
Sure Michael, I think the best place for this then would be with the other aggregation functions in aggregate_by.py . That should tidy things up a bit.

eatyourpeas replied:
There is a folder for functions like this called `general_functions`
Might sit well there.

mbarton commented on epilepsy12/tasks.py:
```diff
@@ -69,10 +73,288 @@ def async_insert_old_pt_data(csv_path: str = "data.csv"):
+
+    final_list = []
+    for kpi in measures:
+        if england_kpi_aggregation[f"{kpi}_total_eligible"] == 0:
```
I think you could avoid another level of nesting here, something like

```python
item = {
    "Country": "England",
    "Measure": kpi,
    "Percentage": 0 if england_kpi_aggregation[f"{kpi}_total_eligible"] == 0 else
        england_kpi_aggregation[f"{kpi}_passed"]
          / england_kpi_aggregation[f"{kpi}_total_eligible"]
          * 100,
    "Numerator": england_kpi_aggregation[f"{kpi}_passed"],
    "Denominator": england_kpi_aggregation[f"{kpi}_total_eligible"],
}
```

pacharanero commented on epilepsy12/tasks.py:
```diff
@@ -69,10 +73,288 @@ def async_insert_old_pt_data(csv_path: str = "data.csv"):
+def download_kpi_summary_as_csv(cohort):
+    """
+    Asynchronous task to pull data from KPIAggregation tables and store as dataframe for export as CSV
+    Accepts cohort as optional param, defaults to 6
```
the default to 6 has been removed as it seemes likely to rot quickly over time

mbarton commented on epilepsy12/tasks.py:
```diff
@@ -69,10 +73,288 @@ def async_insert_old_pt_data(csv_path: str = "data.csv"):
+    # COUNTRY - SHEET 1
+    # create a dataframe with a row for each measure of each country, and a column for each of ["Measure", "Percentage", "Numerator", "Denominator"]
+    
+    CountryKPIAggregation = apps.get_model("epilepsy12", "CountryKPIAggregation")
```
I think this calculation is another good candidate to be a separate function, potentially parameterised with the country?

I was initially a bit confused why this didn't iterate through all the country KPI aggregation models but I think it's because the template we're following just includes England and Wales

dc2007git replied:
That's exactly it - we only calculate England and Wales (not Scotland or NI). Also England and Wales need to be in the same dataframe. Happy to put this into a function though if you think it would be tidier?

eatyourpeas replied:
National covers both England and Wales.

dc2007git replied:
Yep National does indeed. Sheet 1 specifies per Country, so we have to split it up into England and Wales and use the CountryKPIAggregation.

dc2007git committed:
Tidies organisation of create_KPI_aggregation_dataframe function, improves function documentation

dc2007git committed:
Tidies Country KPI aggregation sheet calculations to remvoe excess nesting

mbarton commented:
EDIT: confirmed they do https://github.com/rcpch/rcpch-audit-engine/pull/820/files#diff-22fa9c4678deb8c0f4b25503f6719c4bc756da7e43f0f33d6ad815aee820f6efR809

[from code review in Jitsi] check that the queries set open_access=false to get the latest data

dc2007git committed:
Creates reference dataframe for KPI excel doc and adds sheet

dc2007git committed:
Sets open_access to False in KPI model calls

dc2007git committed:
Removes indexing column in excel doc

dc2007git committed:
Tidies Measure formatting in excel

dc2007git committed:
Adds CountryMeasure column to Country_level sheet

dc2007git committed:
Adds extra string concatenation column to each sheet as requested by E12 team

dc2007git committed:
Updates list of NHS trusts with most recent NHS ODS published data

dc2007git committed:
removes Welsh NHS trust bodies

dc2007git committed:
Removes Liverpool Heart and Chest from trusts.py

mbarton commented:
@dc2007git @eatyourpeas what's left to get this one merged?

eatyourpeas commented:
SOrry this fell off my radar a little. @dc2007git this is your baby. What do you need us to do to get it merged? Do you need either of us to pair with you to finish or are you happy to bring the branch up to date and then merge?

mbarton committed:
Merge branch 'development' into 791-kpi-data-export-specification

mbarton commented:
I've fixed the merge conflicts, it runs locally for me and I'm able to download a report. Investigating the CI errors now.

mbarton commented:
CI is green on a test PR (https://github.com/rcpch/rcpch-audit-engine/pull/897) so it's a problem with this PR specifically.

Reducing the number of tests to run has allowed CI to report an error before it's killed for using too much memory.

https://github.com/rcpch/rcpch-audit-engine/actions/runs/8813952625/job/24192803187?pr=820

```
_ ERROR at setup of test_calculate_kpi_value_counts_queryset_all_levels[EnumAbstractionLevel.TRUST-abstraction_codes1-ods_codes1] _
request = <SubRequest '_django_db_marker' for <Function test_calculate_kpi_value_counts_queryset_all_levels[EnumAbstractionLevel.TRUST-abstraction_codes1-ods_codes1]>>
    @pytest.fixture(autouse=True)
    def _django_db_marker(request: pytest.FixtureRequest) -> None:
        """Implement the django_db marker, internal to pytest-django."""
        marker = request.node.get_closest_marker("django_db")
        if marker:
>           request.getfixturevalue("_django_db_helper")
/usr/local/lib/python3.11/site-packages/pytest_django/plugin.py:532
```

I'm not really sure exactly what's gone wrong. The tests pass locally for me.

I'll try completing wiping my local instance and then running the tests to see if I also get an error. That would isolate it to the setup or migration code.

mbarton commented:
OK that's given me an error running `0005_seed_organisations`:

```
django-1   | INFO [epilepsy12.migrations.0005_seed_organisations:178]  Updating all RCPCH organisations with OPEN UK network relationships... 
django-1   | Traceback (most recent call last):
django-1   |   File "/app/manage.py", line 22, in <module>
django-1   |     main()
django-1   |   File "/app/manage.py", line 18, in main
django-1   |     execute_from_command_line(sys.argv)
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
django-1   |     utility.execute()
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/core/management/__init__.py", line 436, in execute
django-1   |     self.fetch_command(subcommand).run_from_argv(self.argv)
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/core/management/base.py", line 413, in run_from_argv
django-1   |     self.execute(*args, **cmd_options)
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/core/management/base.py", line 459, in execute
django-1   |     output = self.handle(*args, **options)
django-1   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/core/management/base.py", line 107, in wrapper
django-1   |     res = handle_func(*args, **kwargs)
django-1   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/core/management/commands/migrate.py", line 356, in handle
django-1   |     post_migrate_state = executor.migrate(
django-1   |                          ^^^^^^^^^^^^^^^^^
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/db/migrations/executor.py", line 135, in migrate
django-1   |     state = self._migrate_all_forwards(
django-1   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
django-1   |     state = self.apply_migration(
django-1   |             ^^^^^^^^^^^^^^^^^^^^^
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
django-1   |     state = migration.apply(state, schema_editor)
django-1   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/db/migrations/migration.py", line 132, in apply
django-1   |     operation.database_forwards(
django-1   |   File "/usr/local/lib/python3.11/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
django-1   |     self.code(from_state.apps, schema_editor)
django-1   |   File "/app/epilepsy12/migrations/0005_seed_organisations.py", line 199, in seed_organisations
django-1   |     raise Exception(f"{trust_openuk_network['ods trust code']} error")
django-1   | Exception: RC1 error
django-1   |   Applying epilepsy12.0005_seed_organisations...
```

It appears that "Eastern Paediatric Epilepsy Network" has a trust code of "RC1" but there is no corresponding entry in the Trust table.

It is present on `development`: `BEDFORD HOSPITAL NHS TRUST`

https://github.com/rcpch/rcpch-audit-engine/blob/fa3763edb130562a9c84e8a7c69d91b345a0c96a/epilepsy12/constants/trust.py#L1963

but has been removed in this PR. There are a LOT of changes to `trust.py` in this PR. I had thought it was formatting changes but clearly there's also some modifications too.

I'll try and work out which ones have been added/removed/changed and which ones have simply been reformatted. As Amani mentioned in #895 they are able to modify the report to account for recent mergers. I also think it would be clearer to split out those changes from this PR entirely so we have a record of individual modifications or at least remove any incidental formatting changes to the list from this PR

mbarton commented:
OK looking back at the thread in https://github.com/rcpch/rcpch-audit-engine/issues/791 I think the trust changes were done to try and improve the report. Given we have a deadline and will need a few rounds of checking the output I am going to revert the trust changes from this PR and then open any additional ones required separately.

mbarton committed:
Restore trusts.py to the version on development

Fixes an error during seeding where we were missing a trust referenced from an Open UK Network

https://github.com/rcpch/rcpch-audit-engine/pull/820#issuecomment-2074498452

mbarton committed:
Remove unused import

mbarton committed:
Workaround (hopefully) transient signing error with Microsofts ubuntu repos

E: Failed to fetch https://packages.microsoft.com/ubuntu/22.04/prod/dists/jammy/InRelease  Clearsigned file isn't valid, got 'NOSPLIT' (does the network require authentication?)
E: The repository 'https://packages.microsoft.com/ubuntu/22.04/prod jammy InRelease' is no longer signed.

https://github.com/rcpch/rcpch-audit-engine/actions/runs/8814593005/job/24194775367?pr=820

mbarton commented:
CI is green :)

I will merge this PR and send it out to staging. We can make further changes then before testing in production, although I'm happy that this PR should not have any adverse effect as it's purely additive code now the trust changes have gone
